C compilation error with generics and nested if expressions

### Describe the bug

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/bug.01JM1Y22T3STYJDSY1ZAVTAY18.tmp.c:9296: error: expression expected before 'if'
=================================================================
```

### Reproduction Steps

```v
struct Message {
	prio bool
}

struct Throttler[T] {
	max_wait int
mut:
	wait_counter int
	low_prio_requests []T
	high_prio_requests []T
}

fn new_throttler[T](max_wait int) Throttler[T] {
	return Throttler[T] {
		max_wait, 0, []T{cap: 512}, []T{cap: 512}
	}
}

fn pop[T](mut t Throttler[T]) ?T {
	if t.wait_counter >= t.max_wait {
		return if t.high_prio_requests.len > 0 {
			t.high_prio_requests.pop()
		} else {
			t.wait_counter = 0
			if t.low_prio_requests.len > 0 {
				t.low_prio_requests.pop()
			} else {
				none
			}
		}
	} else {
		return none
	}
}

fn main() {
	c := chan Message{}
	mut throttler := new_throttler[Message](10)
	msg := pop(mut throttler) or {
		return
	}

}

```

### Expected Behavior

It should compile

### Current Behavior

```
================== C compilation error (from tcc): ==============
cc: /tmp/v_1000/bug.01JM1Y4BT1ZDYYPF0EE7S1HT25.tmp.c:9283: error: expression expected before 'if'
=================================================================
```

### Possible Solution

_No response_

### Additional Information/Context

_No response_

### V version

V 0.4.9 5be2fca

### Environment details (OS name and version, etc.)

|V full version      |V 0.4.9 5be2fca
|:-------------------|:-------------------
|OS                  |linux, Ubuntu 24.10
|Processor           |16 cpus, 64bit, little endian, AMD Ryzen 7 8840U w/ Radeon 780M Graphics
|Memory              |18.25GB/22.78GB
|                    |
|V executable        |/home/sharp/dev/v/v
|V last modified time|2025-02-13 12:21:16
|                    |
|V home dir          |OK, value: /home/sharp/dev/v
|VMODULES            |OK, value: /home/sharp/.vmodules
|VTMP                |OK, value: /tmp/v_1000
|                    |
|Git version         |git version 2.45.2
|V git status        |weekly.2025.03-35-g5be2fcab-dirty (102 commit(s) behind V master)
|.git/config present |true
|                    |
|cc version          |cc (Ubuntu 14.2.0-4ubuntu2) 14.2.0
|gcc version         |gcc (Ubuntu 14.2.0-4ubuntu2) 14.2.0
|clang version       |Ubuntu clang version 19.1.1 (1ubuntu1)
|tcc version         |tcc version 0.9.28rc 2024-07-31 HEAD@1cee0908 (x86_64 Linux)
|tcc git status      |thirdparty-linux-amd64 0134e9b9
|emcc version        |emcc (Emscripten gcc/clang-like replacement + linker emulating GNU ld) 3.1.6 ()
|glibc version       |ldd (Ubuntu GLIBC 2.40-1ubuntu3.1) 2.40
> [!NOTE]
> You can use the 👍 reaction to increase the issue's priority for developers.
>
> Please note that only the 👍 reaction to the issue itself counts as a vote.
> Other reactions and those to comments will not be taken into account.

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -339,7 +339,6 @@ pub fn gen(files []&ast.File, mut table ast.Table, pref_ &pref.Preferences) GenO
 			label:        'global_cgen'
 		)
 		inner_loop:           unsafe { &ast.empty_stmt }
-		last_if_option_type:  ast.void_type
 		field_data_type:      table.find_type('FieldData')
 		enum_data_type:       table.find_type('EnumData')
 		variant_data_type:    table.find_type('VariantData')
@@ -821,7 +820,6 @@ fn cgen_process_one_file_cb(mut p pool.PoolProcessor, idx int, wid int) &Gen {
 			label:        'cgen_process_one_file_cb idx: ${idx}, wid: ${wid}'
 		)
 		inner_loop:            &ast.empty_stmt
-		last_if_option_type:   ast.void_type
 		field_data_type:       global_g.table.find_type('FieldData')
 		enum_data_type:        global_g.table.find_type('EnumData')
 		variant_data_type:     global_g.table.find_type('VariantData')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -339,6 +339,7 @@ pub fn gen(files []&ast.File, mut table ast.Table, pref_ &pref.Preferences) GenO
 			label:        'global_cgen'
 		)
 		inner_loop:           unsafe { &ast.empty_stmt }
+		last_if_option_type:  ast.void_type
 		field_data_type:      table.find_type('FieldData')
 		enum_data_type:       table.find_type('EnumData')
 		variant_data_type:    table.find_type('VariantData')
@@ -820,6 +821,7 @@ fn cgen_process_one_file_cb(mut p pool.PoolProcessor, idx int, wid int) &Gen {
 			label:        'cgen_process_one_file_cb idx: ${idx}, wid: ${wid}'
 		)
 		inner_loop:            &ast.empty_stmt
+		last_if_option_type:   ast.void_type
 		field_data_type:       global_g.table.find_type('FieldData')
 		enum_data_type:        global_g.table.find_type('EnumData')
 		variant_data_type:     global_g.table.find_type('VariantData')

--- a/vlib/v/gen/c/cgen.v
+++ b/vlib/v/gen/c/cgen.v
@@ -161,6 +161,7 @@ mut:
 	inside_global_decl        bool
 	inside_interface_deref    bool
 	last_tmp_call_var         []string
+	last_if_option_type       ast.Type // stores the expected if type on nested if expr
 	loop_depth                int
 	ternary_names             map[string]string
 	ternary_level_names       map[string][]string
@@ -2068,7 +2069,8 @@ fn (mut g Gen) stmts_with_tmp_var(stmts []ast.Stmt, tmp_var string) bool {
 								styp = 'f64'
 							}
 						}
-						if stmt.typ.has_flag(.option) {
+						if stmt.typ.has_flag(.option)
+							|| (g.inside_if_option && stmt.expr is ast.IfExpr) {
 							g.writeln('')
 							g.write('${tmp_var} = ')
 							g.expr(stmt.expr)
@@ -2082,7 +2084,6 @@ fn (mut g Gen) stmts_with_tmp_var(stmts []ast.Stmt, tmp_var string) bool {
 							}
 							styp = g.base_type(ret_typ)
 							if stmt.expr is ast.CallExpr && stmt.expr.is_noreturn {
-								g.expr(stmt.expr)
 								g.writeln(';')
 							} else {
 								g.write('_option_ok(&(${styp}[]) { ')

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -185,15 +185,26 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 	// easier to use a temp var, than do C tricks with commas, introduce special vars etc
 	// (as it used to be done).
 	// Always use this in -autofree, since ?: can have tmp expressions that have to be freed.
-	needs_tmp_var := g.need_tmp_var_in_if(node)
+	needs_tmp_var := g.inside_if_option || g.need_tmp_var_in_if(node)
 	needs_conds_order := g.needs_conds_order(node)
-	tmp := if node.typ != ast.void_type && needs_tmp_var { g.new_tmp_var() } else { '' }
+	tmp := if g.inside_if_option || (node.typ != ast.void_type && needs_tmp_var) {
+		g.new_tmp_var()
+	} else {
+		''
+	}
 	mut cur_line := ''
 	mut raw_state := false
 	if needs_tmp_var {
 		mut styp := g.styp(node.typ)
-		if node.typ.has_flag(.option) {
+		if g.inside_if_option || node.typ.has_flag(.option) {
 			raw_state = g.inside_if_option
+			if node.typ != ast.void_type {
+				tmp_if_option_type := g.last_if_option_type
+				g.last_if_option_type = node.typ
+				defer {
+					g.last_if_option_type = tmp_if_option_type
+				}
+			}
 			defer {
 				g.inside_if_option = raw_state
 			}
@@ -210,7 +221,13 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 		cur_line = g.go_before_last_stmt()
 		g.empty_line = true
 		if tmp != '' {
-			g.writeln('${styp} ${tmp}; /* if prepend */')
+			if node.typ == ast.void_type && g.last_if_option_type != 0 {
+				// nested if on return stmt
+				g.write2(g.styp(g.last_if_option_type), ' ')
+			} else {
+				g.write('${styp} ')
+			}
+			g.writeln('${tmp}; /* if prepend */')
 		}
 		if g.infix_left_var_name.len > 0 {
 			g.writeln('if (${g.infix_left_var_name}) {')

--- a/vlib/v/gen/c/if.v
+++ b/vlib/v/gen/c/if.v
@@ -194,12 +194,12 @@ fn (mut g Gen) if_expr(node ast.IfExpr) {
 	}
 	mut cur_line := ''
 	mut raw_state := false
+	tmp_if_option_type := g.last_if_option_type
 	if needs_tmp_var {
 		mut styp := g.styp(node.typ)
 		if g.inside_if_option || node.typ.has_flag(.option) {
 			raw_state = g.inside_if_option
 			if node.typ != ast.void_type {
-				tmp_if_option_type := g.last_if_option_type
 				g.last_if_option_type = node.typ
 				defer {
 					g.last_if_option_type = tmp_if_option_type

--- a/vlib/v/tests/nested_if_return_test.v
+++ b/vlib/v/tests/nested_if_return_test.v
@@ -1,0 +1,41 @@
+struct Message {
+	prio bool
+}
+
+struct Throttler[T] {
+	max_wait int
+mut:
+	wait_counter       int
+	low_prio_requests  []T
+	high_prio_requests []T
+}
+
+fn new_throttler[T](max_wait int) Throttler[T] {
+	return Throttler[T]{max_wait, 0, []T{len: 1, init: T{}}, []T{len: 1, init: T{}}}
+}
+
+fn pop[T](mut t Throttler[T]) ?T {
+	if t.wait_counter <= t.max_wait {
+		return if t.high_prio_requests.len > 10 {
+			t.high_prio_requests.pop()
+		} else {
+			t.wait_counter = 0
+			if t.low_prio_requests.len > 0 {
+				t.low_prio_requests.pop()
+			} else {
+				none
+			}
+		}
+	} else {
+		return none
+	}
+}
+
+fn test_main() {
+	c := chan Message{}
+	mut throttler := new_throttler[Message](10)
+	msg := pop(mut throttler) or { return }
+	assert msg.str() == 'Message{
+    prio: false
+}'
+}

